### PR TITLE
fix(expo, asset, cli)!: unify asset hashing

### PR DESCRIPTION
**Pull request by EvanBacon**
# Why

Convert the `../` segments of the server URL to `_` to support monorepos. This same transformation takes place in `AssetSourceResolver.web` (expo-assets, expo-image) and `persistMetroAssets` of Expo CLI, this originally came from the [Metro opinion](https://github.com/react-native-community/cli/blob/2204d357379e2067cebe2791e90388f7e97fc5f5/packages/cli-plugin-metro/src/commands/bundle/getAssetDestPathIOS.ts#L19C5-L19C10).

The purpose is to ensure no URL like `/foo/../bar.png` is requested, as the result would be `/bar.png` which wouldn't work. The of how to handle this didn't come from us, but we do need to ensure it's unified. At a high-level, this does prevent the usage of certain files, as `/foo/../bar.png` will be `/foo/_bar.png`, meaning a file named `/foo/_bar.png` cannot also exist. This logic, while applied at runtime, is actually only valid for production exports as we don't move or alias files in development. The only way to have valid development files is to ensure `../` never appears in the URL, i.e. by using `unstable_serverRoot`.

- Drop legacy `expo/tools/hashAssetFiles.js` in favor of `expo-asset` version.
- Unify runtime logic of asset file loading for monorepos.
- Split out of https://github.com/expo/expo/pull/23911

# How

# Test Plan

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-asset/tools/hashAssetFiles.js [hunk 1]
```diff
@@ -3,8 +3,19 @@
...
+  return Promise.all(asset.files.map(md5File)).then((hashes) => {
     asset.fileHashes = hashes;
+
+    // Convert the `../` segments of the server URL to `_` to support monorepos.
+    // This same transformation takes place in `AssetSourceResolver.web` (expo-assets, expo-image) and `persistMetroAssets` of Expo CLI,
+    // this originally came from the Metro opinion https://github.com/react-native-community/cli/blob/2204d357379e2067cebe2791e90388f7e97fc5f5/packages/cli-plugin-metro/src/commands/bundle/getAssetDestPathIOS.ts#L19C5-L19C10
+    if (asset.httpServerLocation.includes('?export_path=')) {
+      asset.httpServerLocation = asset.httpServerLocation
+        .match(/\?export_path=(.*)/)[1]
+        .replace(/\.\.\//g, '_');
```
Comment (ide) [inline packages/expo-asset/tools/hashAssetFiles.js:15]: This logic seems fragile: consider a directory named `x..`

**Review by ide — CHANGES_REQUESTED** [review]

File: packages/expo-asset/tools/hashAssetFiles.js [hunk 2]
```diff
@@ -3,8 +3,19 @@
...
+    }
+    // asset.httpServerLocation = asset.httpServerLocation.replace(/\.\.\//g, '_');
```
Comment (EvanBacon) [inline packages/expo-asset/tools/hashAssetFiles.js:17]:
```suggestion
```

File: packages/expo-asset/tools/hashAssetFiles.js (hunk 1 above)
Comment (ide) [reply packages/expo-asset/tools/hashAssetFiles.js:15]: ok, I suppose buggy in a consistent way is better than being sometimes correct

**Review by ide — APPROVED** [review]

**Comment by RohovDmytro** [comment]
@EvanBacon, is it expected to be used with expo v49?

Changelog says its a breaking change and my v49 app  stopped working.

**Pull request merged**